File: sdk/typescript/src/protocol.ts
/**
 * Shared protocol types for the Orbit SDK.
 *
 * The SDK wraps the `orbit` CLI. The event names (`item.completed`,
 * `turn.completed`) and the `TurnResult` shape follow the SDK README; they are
 * mapped from the CLI's `--output-format json --stream` JSONL output.
 */

export type JsonPrimitive = string | number | boolean | null;
export type JsonValue = JsonPrimitive | JsonValue[] | { [key: string]: JsonValue };

/** A single item produced during a turn. */
export type ThreadItem =
  | { type: "text"; text: string }
  | { type: "tool_use"; name: string; input: string }
  | { type: "tool_result"; content: string }
  | { type: "image"; path?: string; url?: string };

/** Token usage reported by the CLI. */
export interface Usage {
  input_tokens: number;
  output_tokens: number;
  cache_creation_input_tokens: number;
  cache_read_input_tokens: number;
}

/** The buffered result of a completed turn. */
export interface TurnResult {
  finalResponse: string;
  items: ThreadItem[];
  usage?: Usage;
}

/**
 * Structured events emitted while a turn runs. `runStreamed` yields these;
 * `run` buffers them into a {@link TurnResult}.
 */
export type OrbitEvent =
  | { type: "turn.started" }
  | { type: "item.completed"; item: ThreadItem }
  | {
      type: "turn.completed";
      finalResponse: string;
      usage?: Usage;
      sessionId?: string;
    }
  | { type: "turn.failed"; error: string };

/** Structured input entry accepted by {@link Thread.run}. */
export type InputEntry =
  | { type: "text"; text: string }
  | { type: "local_image"; path: string };

/** A turn can be started with a plain prompt string or structured entries. */
export type ThreadInput = string | InputEntry[];

/** Options for constructing the {@link Orbit} client. */
export interface OrbitOptions {
  /** Environment passed to the spawned CLI. Merged over `process.env`. */
  env?: Record<string, string>;
  /** Base URL the CLI should use; passed as a `frontal_base_url` config override. */
  baseUrl?: string;
  /** Global CLI `--config` overrides. */
  config?: Record<string, JsonValue>;
  /** Path to the `orbit` CLI binary. Defaults to `"orbit"`. */
  command?: string;
}

/** Options for starting or resuming a {@link Thread}. */
export interface ThreadOptions {
  /** Run the CLI with this working directory. */
  workingDirectory?: string;
  /** Skip the CLI's Git repository check. */
  skipGitRepoCheck?: boolean;
  /** Provider override (e.g. `anthropic`, `frontal`). */
  provider?: string;
  /** Model override (e.g. `opus`, `claude-opus-4-6`). */
  model?: string;
  /** Permission mode override. */
  permissionMode?: string;
  /** Thread-level CLI `--config` overrides (take precedence over global). */
  config?: Record<string, JsonValue>;
}

/** Per-turn options for {@link Thread.run} / {@link Thread.runStreamed}. */
export interface ThreadRunOptions {
  /** JSON schema the agent should conform its response to. */
  outputSchema?: JsonValue;
  /** Run-level CLI `--config` overrides (highest precedence). */
  config?: Record<string, JsonValue>;
}

File: sdk/rust/src/protocol.rs
//! Shared protocol types for the Orbit SDK.
//!
//! The SDK wraps the `orbit` CLI. The event names (`item.completed`,
//! `turn.completed`) and the [`TurnResult`] shape follow the SDK README and are
//! mapped from the CLI's `--output-format json --stream` JSONL output.

/// A single item produced during a turn.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ThreadItem {
  Text {
    text: String,
  },
  ToolUse {
    name: String,
    input: String,
  },
  ToolResult {
    content: String,
  },
  Image {
    path: Option<String>,
    url: Option<String>,
  },
}

/// Token usage reported by the CLI.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Usage {
  pub input_tokens: u64,
  pub output_tokens: u64,
  pub cache_creation_input_tokens: u64,
  pub cache_read_input_tokens: u64,
}

/// The buffered result of a completed turn.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct TurnResult {
  pub final_response: String,
  pub items: Vec<ThreadItem>,
  pub usage: Option<Usage>,
}

/// Structured events emitted while a turn runs.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum OrbitEvent {
  TurnStarted,
  ItemCompleted {
    item: ThreadItem,
  },
  TurnCompleted {
    final_response: String,
    usage: Option<Usage>,
    session_id: Option<String>,
  },
  TurnFailed {
    error: String,
  },
}

/// Structured input entry accepted by [`Thread::run`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum InputEntry {
  Text { text: String },
  LocalImage { path: String },
}

/// A turn can be started with a plain prompt string or structured entries.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ThreadInput {
  Text(String),
  Entries(Vec<InputEntry>),
}

/// Options for constructing the [`Orbit`] client.
#[derive(Debug, Clone, Default)]
pub struct OrbitOptions {
  /// Environment passed to the spawned CLI. Merged over the ambient env.
  pub env: Option<std::collections::HashMap<String, String>>,
  /// Base URL the CLI should use; passed as a `frontal_base_url` config
  /// override.
  pub base_url: Option<String>,
  /// Global CLI `--config` overrides.
  pub config: Option<serde_json::Value>,
  /// Path to the `orbit` CLI binary. Defaults to `"orbit"`.
  pub command: Option<String>,
}

/// Options for starting or resuming a [`Thread`].
#[derive(Debug, Clone, Default)]
pub struct ThreadOptions {
  /// Run the CLI with this working directory.
  pub working_directory: Option<std::path::PathBuf>,
  /// Skip the CLI's Git repository check.
  pub skip_git_repo_check: bool,
  /// Provider override (e.g. `anthropic`, `frontal`).
  pub provider: Option<String>,
  /// Model override (e.g. `opus`, `claude-opus-4-6`).
  pub model: Option<String>,
  /// Permission mode override.
  pub permission_mode: Option<String>,
  /// Thread-level CLI `--config` overrides (precedence over global).
  pub config: Option<serde_json::Value>,
}

/// Per-turn options for [`Thread::run`].
#[derive(Debug, Clone, Default)]
pub struct ThreadRunOptions {
  /// JSON schema the agent should conform its response to.
  pub output_schema: Option<serde_json::Value>,
  /// Run-level CLI `--config` overrides (highest precedence).
  pub config: Option<serde_json::Value>,
}

impl ThreadInput {
  /// Build the CLI prompt string and the list of image paths.
  pub fn to_prompt_and_images(&self) -> (String, Vec<String>) {
    match self {
      ThreadInput::Text(text) => (text.clone(), Vec::new()),
      ThreadInput::Entries(entries) => {
        let mut texts = Vec::new();
        let mut images = Vec::new();
        for entry in entries {
          match entry {
            InputEntry::Text { text } => texts.push(text.clone()),
            InputEntry::LocalImage { path } => images.push(path.clone()),
          }
        }
        (texts.join("\n"), images)
      }
    }
  }
}
